components/ingest-service/pipeline/publisher/msg_distributor.go

```diff
@@ -0,0 +1,90 @@
+package publisher
+
+import (
+	"time"
+
+	"github.com/chef/automate/components/ingest-service/pipeline/message"
+	log "github.com/sirupsen/logrus"
+)
+
+// BuildMsgDistributor a message distributor that sends all messages to the first child pipe until it is
+// full then start sending messages to the next child pipe in the line. When all the pipes are full the thread waits and
+// then checks to pipes again until it can place the message in a child pipe.
+func BuildMsgDistributor(childPipeBuilder message.ChefRunPipe, numProcessors, childPipeInboxSize int) message.ChefRunPipe {
+	return func(in <-chan message.ChefRun) <-chan message.ChefRun {
+		return msgDistributor(in, childPipeBuilder, numProcessors, childPipeInboxSize)
+	}
+}
+
+func msgDistributor(in <-chan message.ChefRun,
+	childPipeBuilder message.ChefRunPipe, numProcessors, childPipeInboxSize int) <-chan message.ChefRun {
+	pipeInChannels, out := buildChildPipes(childPipeBuilder, numProcessors, childPipeInboxSize)
+
+	go func() {
+		for msg := range in {
+			sendMessage(pipeInChannels, msg)
+		}
+		close(out)
```

Comment:
Im not so sure about it but, should this close func call belongs in a defer statement? 🤔 (I'm just starting to think about what happens when one of these goroutines fail for any reason.)

Reply:
I think you might be correct. I don't want to make this change right now though, because this has already been thoroughly tested. This is a problem with all the processors in the pipeline and should be tested separately. 

Good catch!